### ISSUE TITLE
bracketSameLine sometimes breaks to a new line when it shouldn't

### ISSUE THREAD

**Issue by karptonite** [prettier/prettier#11484]
**Prettier 2.4.0**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEACAPAZwLYEMA2+qAlgCYC8AOiABZz4AOcATiadamPrpplSADMIzbAFp4ADxipJMUdgCu8diAB8AQVT5oAc2KYaqBsThg4qCAJlwpqOszjoA9DgL5VlKJ-QNPqVABUUDoAkgL8MDT6AHQMuA6w0aRwmGDMxAwwwpioAGS5MlGYsfEIMEkpaRlZzMX4CDqR1H6c3Lz8DBCYcthwMLjUHlD+WHFew-7+XDx81J3d8n24APrLxPDYzRP+QToAYsL89dLJqemZwhZWkTFxCeWnVRe1W5OoGlAQkSxGJmZX1ls9kcLjGQ2cDFUIAANCAIJliNBMMhQPFmBAAO4ABVKUGRKAIGNwAE9kbCAEbMXBgADWfQAyrhegAZYhQODIGDMBRwClU2kMuJgNk6Tnc3kgODYclwUjJUjM3DBBS4HRwA4iXAwGAi5AgXBKCAw2gwbD4ADqUXgmCFcHpCEw62IADd1sS9WBeMa2ZgWDAsVSdHhkAICL7YQArTASeki+oARQUXw5SFD+HDIDuvuYepopvwxoY6Vg5rIkWQAA4AAywotdODmqkMPVFlIsZ0c2EARyT8AD8Px+swonZstlxocPeIDgDquDqbDEt92GIYp5sMdwQTvZTaYz-XJpdI5aQACZYVzcMR8CKAMIQbDzyWYACsxoUvoAKrhyfi9xLnR5EIoGSWB6SeGB1BA+kYGJeoQ0XABfRCgA)

```sh
--parser html
--bracket-same-line
```

**Input:**

```html
 <small id="helper id" class="form-text text-muted">A longish piece of text here</small>

<p
  *ngIf="this.parent.descriptors && this.parent.descriptors.length"
  class="post-meta">
  <span
     class="post-meta__item"
    *ngFor="let descriptor of this.parent.descriptors"
    >Another piece of text here</span>
</p>
```

**Output:**

```html
<small id="helper id" class="form-text text-muted"
  >A longish piece of text here</small
>

<p
  *ngIf="this.parent.descriptors && this.parent.descriptors.length"
  class="post-meta">
  <span
    class="post-meta__item"
    *ngFor="let descriptor of this.parent.descriptors"
    >Another piece of text here</span
  >
</p>

```

**Expected behavior:**
```html
<small id="helper id" class="form-text text-muted"
  >A longish piece of text here</small>

<p
  *ngIf="this.parent.descriptors && this.parent.descriptors.length"
  class="post-meta">
  <span
    class="post-meta__item"
    *ngFor="let descriptor of this.parent.descriptors"
    >Another piece of text here</span>
</p>

```

In other words, it looks like the `bracket-same-line` option might get a bit confused when closing white-space-sensitive elements that are inside non-white-space-sensitive elements, or not inside of other elements. It seems like in both cases, the closing bracket for the closing tag should be on the same line.

**Comment by karptonite** [comment]
This may be of interest to @kurtztech. (Thanks for all your work on this, by the way. Having started trying myself, and failing, I have a sense of what was involved.)

**Comment by pcjmfranken** [comment]
Here's another example of where it goes wrong in `html`:

[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeADgPgEIBsCGARgcfgASoDWhAJplbZgMIwBOuqA9NXQNQN0AygAsAlgDMYXHpn4yAStMZK6AOi5YQAGhAR0MUdADOyUPlasIAdwAK5hCZT5cV-AE8TOwq3xhKcGEF8AFs4ABlRKDhkcWcjOC8fPwDBdF9IgHNkNgBXBJA4YMI4GhoSsPwoDJz8DLgAMQhWYPwYAyrkEHwcmAhtEGEYYNwAdTF4IzSwOEEHUQMAN3m3TrAjTxBI+NYYGx8Mlpi4-IArIwAPQUzcOABFHIh4I9x4nTTWbc7B4f70VkiYCNRDQYMJkAAOAAMb0s8RGPnQnT+cG2C2iOgAjg94Hs9I4ukYALRREolfqsOBY0QUva1Q5IWIvfLxYKibKsPI6IzXO7Y6IM446GBEIEgsFIABMQp8olwmSYEGC9IKRgArP0cvEACpERyM14gBZ5ACSUDKsEEYH++gAgmbBDA3DdnvEAL6uoA)

Input:
```html
<p>Blablabla <kbd><kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>R</kbd></kbd>.</p>
```

Output:
```html
<p>
  Blablabla <kbd><kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>R</kbd></kbd
  >.
</p>
```

Expected:
```html
<p>
  Blablabla <kbd><kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>R</kbd></kbd>.
</p>
```

FYI The expected behaviour is observed when the closing period is not present.

**Comment by karptonite** [comment]
This may be related, but isn't exactly the same issue, since in your case, there is text where the whitespace--or lack thereof--is meaningful, following that last kbd element.

**Comment by karptonite** [comment]
I see now why this has been marked "needs discussion" and not "bug". While the description of "bracket same line" option says that it applies to multi-line HTML **elements**, the example shows a multi-line HTML **tag**, and only shows an opening tag as an example, so it is a bit vague.

I suggest that it is in the spirit of the option for this to also apply to the closing tag of a multi-line HTML element, as shown in the examples above. The overall goal of the option seems to be to avoid having an isolated `>` on a line when possible.

**Comment by radiosilence** [comment]
It seems that whilst `jsxBracketSameLine` is deprecated in favour of `bracketSameLine`, they do not actually do the same thing, so I would say this is a bug.

With `bracketSameLine`:

```tsx
            <FormGroup
              label="County"
              htmlFor="delivery.county"
              required={needsShipping}
              error={errors.delivery?.county}
            >
              <Input
                id="delivery.county"
                placeholder="County"
                error={!!errors.delivery?.county}
                {...register('delivery.county', { required: needsShipping })}
              />
            </FormGroup>
```

With `jsxBracketSameLine`:

```tsx
            <FormGroup
              label="County"
              htmlFor="delivery.county"
              required={needsShipping}
              error={errors.delivery?.county}>
              <Input
                id="delivery.county"
                placeholder="County"
                error={!!errors.delivery?.county}
                {...register('delivery.county', { required: needsShipping })}
              />
            </FormGroup>
```

**Comment by TechPhil** [comment]
I'm in agreement on this - if the purpose is to remove 'floating' brackets, then this should also apply to closing tags.

**Comment by andretf** [comment]
I'm too experiencing this weird output. Although I want to break lines with brackets instead, but it looks like it's the same behavior. I'm forced to add an extra white space to force closing tag goes to the next line, which is less weird. But sill, the closing bracket of the opening tag is indented differently if there is not a <text node> before tag content:

https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAeeBbADgGwIbwB8AOlAARmoAmAlgG4nkWW0OnPPX2McddtNenVjyG9+oseIDOWPFHZSxYfNOkBeYiDzSqOvbv1l5VY1AgwAFnABOZAEYBXKGEtkIAMzIqd0uNK1FJV46AFoaD00QaQgMOCsaKABzLUlgsmIYBOkyGhzIGxs4MBgg9NQAell5NODUaoVBdIofNSj9DsNdM1N5C2s7Jxc3T29VPwCQMuawiKiYuITk1LyybO8IQuKYAH5ppUqGxn2perlG5o5ZyK1OvUCm5taNLQX4y0SktcsCMiwbRIwADuNCoVhWOXWBSKJR2h3OtQODROSmu7S69ymj3SzyiIKsZDgAA8YDY8GQgR94GRqmA4BDvqtodsdpQqgjSCihGd5FyxGjbhi8A9Lsxca9Yu9Pt9fhg8ABrfy5GB-Xx-AGwEFgywMqGbGG7eE1Pl8ZHY4Li7RCjE9Mz9WwOZyudxeZ7+EWisgC6KSpYpLGiwirPVbWFGqCI05mz1e8I3K0GTEm3iW8k4Cwusa+JX4tzE0nkyk0am0+kBy5ByEffL6llso7JyoiPlN7go1sCbkVZtNDuiPukSqYXAEOCMEAAGhAECwMBo0GkyFAeEKECBAAUVwhFyg8DggXgAJ6Lqf2MlgRUwADKeDiABlEnBkB4935T+fL1e5GBPshSY44CnOAMHsOAqCoMC73kJJHDwJI4AAMU2OUsl-XdHBgCBJxASwYAwHAAHUqX8b84Cvbdi3oYtD2QaJSRoEpsMSPwbBgdcySSOVn1fQCQAAK2kIkr0+HA4AARUcCwnyQF8cDfEA5BsFjaLoADsP+QECNBKxkAADgABinf4ID8AiySwWj-n8Ww6CfKcAEdJPgdiZx3K1QigOAwLA7CikcmginYuCuJknipz8DAaD-GwAPCkTxKc6TZPkmA8HsLTtWQAAmKcCxoHBPgAYViEKQH8ABWbDHD8AAVNKd2S3jVLgABJKAINgK8wABWcAEF2qvGBD1E7i5LgABfcagA

**Comment by karptonite** [comment]
I really think a new option is the wrong approach here. I can't imagine a case where someone would prefer to avoid the stranded ">" in opening but not closing tags, and of course, prettier is not in the business of adding unnecessary options.

**Comment by mzagaja** [comment]
As a newer prettier user who has been experimenting with this tool, I agree that the current behavior is not expected. From a developer experience standpoint I do not want stranded `>` characters on their own line. The stranded `>` characters make the readability of the HTML subjectively worse. For the most part I do not plan to adopt Prettier for HTML documents unless there is a way to avoid stranded `>`.

**Comment by 1nfected** [comment]
I came here to report/request exactly this thing.

Seeing the '>' dangling there all by itself makes me sad!

Formatting is supposed to make reading code **easier**! And this weird little behaviour actually does the opposite.
All these '>' characters dangling all by themselves cause a distraction.

**Comment by mgenware** [comment]
I feel like `bracketSameLine` is doing the opposite of what I wish.

What it does now:
```html
  <my-comp
    attr
    attr2="sdfjhhiewufwheif"
    attr3="sdfjhhiewufwheif"
    attr4="sdfjhhiewufwheif"
    >text</my-comp
  >
```

What it should be:
```html
  <my-comp
    attr
    attr2="sdfjhhiewufwheif"
    attr3="sdfjhhiewufwheif"
    attr4="sdfjhhiewufwheif"
  >text</my-comp>
```

**Comment by Ornhoj** [comment]
I must say I've never seen a need for the closing tags ending bracket being on a new line. The cases I've seen which I would expect this property to handle would be among the following. (At least these are the outputs I've seen in different codebases)

**bracketSameLine: false**
A
```
<MyComponent
    attr="value"
    attr2="value"
/>
```
B
```
<MyComponent
    attr="value"
    attr2="value"
>Text</MyComponent>
```
C
```
<MyComponent
    attr="value"
    attr2="value"
>
    Text
</MyComponent>
```

**bracketSameLine: true**
A
```
<MyComponent
    attr="value"
    attr2="value" />
```

B
```
<MyComponent
    attr="value"
    attr2="value">Text</MyComponent>
```

C
```
<MyComponent
    attr="value"
    attr2="value">
    Text
</MyComponent>
```

Personally I'm in favour of variant B - the introduced issue is when/if the behaviour in variant C should be addressed with a new property, be affected by `bracketSameLine` or simply be dictated by `printWidth` and the already implemented behaviour of Prettier

**Comment by MiguelX413** [comment]
This is the only thing in the way of me using prettier.

**Comment by devinci** [comment]
Currently experiencing this as well, thought there was something up with my config as I would have expected to render as the following:

`bracketSameLine: true` 
`singleAttributePerLine: true`
```
<MyComponent
    attr="value"
    attr2="value">
    Text
</MyComponent>
```

`bracketSameLine: true` 
`singleAttributePerLine: false`
```
<MyComponent attr="value" attr2="value">
    Text
</MyComponent>
```
`bracketSameLine: false` 
`singleAttributePerLine: false`
```
<MyComponent attr="value" attr2="value"
>
    Text
</MyComponent
>
```
`bracketSameLine: false` 
`singleAttributePerLine: true`
```
<MyComponent
    attr="value"
    attr2="value"
    >
    Text
</MyComponent
>
```

But to be honest, people are so pedantic that it's probably going to be easier to have a separate config option for closing bracket: 

`bracketSameLine:  ["opening": true, "closing": true] `

**Comment by juliuslipp** [comment]
Please fix the floating tag!!

**Comment by kachidk** [comment]
Try this!

```json
{
  "bracketSameLine": true,
  "bracketSpacing": true,
  "htmlWhitespaceSensitivity": "ignore"
}
```

**Comment by HyperLife1119** [comment]
2023, any updates here?

**Comment by KyTiXo** [comment]
Bumping for an update <3

**Comment by mayurarx** [comment]
```
{
  "jsxBracketSameLine": true
}
```

This still works in a .prettierrc file, using vscode

**Comment by prince272** [comment]
> Bumping for an update <3. +1 for `bracketSameLine: ["opening": true, "closing": true]`

How is that going?

**Comment by NdagiStanley** [comment]
Any update on this?

**Comment by AmelianceSkyMusic** [comment]
Any update on this?

**Comment by dionatan-g** [comment]
I believe there's something weird happening with this config.
If you run from the cli, the bracketSameLine works, but from the extension it doesn't.
So I tried to uninstall > install the extension and it worked, now the closing tag is at the same line.
Maybe the prettier extension got lost when I changed the prettierrc config?

**Comment by pedram6195** [comment]
> I believe there's something weird happening with this config. If you run from the cli, the bracketSameLine works, but from the extension it doesn't. So I tried to uninstall > install the extension and it worked, now the closing tag is on the same line. Maybe the prettier extension got lost when I changed the prettierrc config?
> 
> prettier: v3.0.0 prettier-extension: v9.19.0

I had the same issue, reinstalled the extension and it worked.

**Comment by gabrielmarins0** [comment]
Ativando "bracketSameLine" nas configurações padrão:

![image](https://github.com/prettier/prettier/assets/123405963/ad91f77b-583f-4871-88db-56c9c89c48cb)

Resultado:
![image](https://github.com/prettier/prettier/assets/123405963/27aa901a-95d0-4878-a2a2-afd27b7b808a)

Colocando um arquivo .prettierrc.josn na raiz do projeto com  o seguinte código:

`{
  "bracketSameLine": true
}`

Resultado:
![image](https://github.com/prettier/prettier/assets/123405963/7189274f-1e7b-4bee-a40b-cb6ce0f0bf76)

**Comment by tomasdev** [comment]
Can someone confirm whether we'll get https://github.com/prettier/prettier/issues/11484#issuecomment-1431701319 or not in a future update?

**Comment by jakejohns** [comment]
Note that this strange and undesirable formatting, is not ***only*** an eyesore, but seems to ***break*** various vim plugins that deal with surrounding tags in certian contexts, eg. [tpope/vim-surround][tpope-surroud], [kylechui/nvim-surround][nvim-surround] 

[nvim-surround]: https://github.com/kylechui/nvim-surround 
[tpope-surroud]: https://github.com/tpope/vim-surround

```html
  <a
    href="really-long/lorenipsumdollarsitamet/foo/bar/baz/bing/bloop/blop/bling"
    >foo</a
  >
```

This formatting leads to unexpected, incorrect, and inconsistent behavior between the mentioned vim plugins.

**Comment by DavidPlonsky** [comment]
Caution, maybe bad english ;)

Uninstalled > reinstalled - does not work.

set this on user and workspace also didn't work.
``` json
{
	"prettier.bracketSameLine": true,
	"prettier.bracketSpacing": true,
	"prettier.htmlWhitespaceSensitivity": "ignore",
	"prettier.jsxBracketSameLine": true,
}
```


result on formatting:

``` vue
<v-col class="flex-grow-0">
	<v-btn
		prepend-icon="mdi-pen"
		@click="on_select()"
		>select entry</v-btn
	>
</v-col>
```

so is this still a bug that isn't fixed or i am the one who doesn't understand this? This is nearly unreadable if we got a file with hundreds of lines.

**Comment by digeomel** [comment]
> Caution, maybe bad english ;)
> 
> Uninstalled > reinstalled - does not work.
> 
> set this on user and workspace also didn't work.
> 
> {
> 	"prettier.bracketSameLine": true,
> 	"prettier.bracketSpacing": true,
> 	"prettier.htmlWhitespaceSensitivity": "ignore",
> 	"prettier.jsxBracketSameLine": true,
> }
> 
> result on formatting:
> 
> <v-col class="flex-grow-0">
> 	<v-btn
> 		prepend-icon="mdi-pen"
> 		@click="on_select()"
> 		>select entry</v-btn
> 	>
> </v-col>
> 
> so is this still a bug that isn't fixed or i am the one who doesn't understand this? This is nearly unreadable if we got a file with hundreds of lines.

You don't mention which version of Prettier you use. With Prettier 3.6.2 and this `.prettierrc` we have no such issues:

```json
{
    "trailingComma": "es5",
    "tabWidth": 4,
    "useTabs": false,
    "singleQuote": true,
    "printWidth": 140,
    "semi": true,
    "bracketSpacing": true,
    "bracketSameLine": true,
    "arrowParens": "always",
    "quoteProps": "preserve",
    "plugins": ["prettier-plugin-organize-imports"]
}
```
If you're using vscode I assume you also have the right extensions installed.

**Comment by ConradGoodwin** [comment]
I would like to add to this. I am using the following:

VSCode: 1.103.2 (on macOS)
Prettier: 11.0.0 (through VSCode)

`.prettierrc` file in the same folder as my .html file:

```
 {
     "trailingComma": "es5",
     "tabWidth": 4,
     "useTabs": false,
     "singleQuote": true,
     "printWidth": 140,
     "semi": true,
     "bracketSpacing": true,
     "bracketSameLine": true,
     "arrowParens": "always",
     "quoteProps": "preserve",
     "plugins": ["prettier-plugin-organize-imports"]
 }
```

Attempting to use format document on:
`<p style="text-align: justify;"><a class="instructure_file_link instructure_scribd_file inline_disabled" title="Questions 1.pdf" href="/courses/41359/files/12273791?wrap=1" target="_blank" rel="noopener noreferrer">Available here</a> (as .pdf)</p>`

Results in:
```
<p style="text-align: justify">
    <a
        class="instructure_file_link instructure_scribd_file inline_disabled"
        title="Questions 1.pdf"
        href="/courses/41359/files/12273791?wrap=1"
        target="_blank"
        rel="noopener noreferrer"
        >Available here</a
    >
    (as .pdf)
</p>
```

As opposed to the expected behaviour:
```
<p style="text-align: justify">
    <a
        class="instructure_file_link instructure_scribd_file inline_disabled"
        title="Questions 1.pdf"
        href="/courses/41359/files/12273791?wrap=1"
        target="_blank"
        rel="noopener noreferrer"
        >Available here
    </a>
    (as .pdf)
</p>
```

The daft portal I'm required to use does not interpret it as a properly closed `</a>`.

I had a bit of a muck around on the Prettier playground, and it gives exactly the same issue. Somewhat bizarrely, simply clicking "require pragma" fixes it.

Furthermore, if I instead use:
`<p style="text-align: justify;"><a class="instructure_file_link instructure_scribd_file inline_disabled" title="Questions 1.pdf" href="/courses/41359/files/12273791?wrap=1" target="_blank" rel="noopener noreferrer">Available here </a> (as .pdf)</p>`

Where there is a single space before the `</a>`, it gives the expected behaviour and puts the whole `</a>` on a new line. But then I have a space after "Available here", which is just underlined white space because it's a link.

Of course, adding `"htmlWhitespaceSensitivity": "ignore"` does make it work, but then that brings other issues in elsewhere.

**Comment by digeomel** [comment]
@ConradGoodwin with your code and these (i.e. my) settings, I also get the same results. But your "expected behaviour" is wrong. What you would expect is this:

```html
<p style="text-align: justify">
    <a
        class="instructure_file_link instructure_scribd_file inline_disabled"
        title="Questions 1.pdf"
        href="/courses/41359/files/12273791?wrap=1"
        target="_blank"
        rel="noopener noreferrer"
        >Available here</a>
    (as .pdf)
</p>
```
In other words, for whitespace issues, the closing `</a>` tag will never be alone on a different line.
Anyway, if prettier doesn't work for you I suggest you try other, more modern alternatives, there are new kids on the block,
